Release/src/http/client/winhttppal/winhttppal.cpp

```diff
@@ -0,0 +1,3644 @@
+        return 0;
+    default: /* in case a new one is introduced to shock us */
+        TRACE_VERBOSE("%-35s:%-8d:%-16p type:%d\n", __func__, __LINE__, (void*)userp, type);
+        return 0;
+
+    case CURLINFO_HEADER_IN:
+        text = "=> Receive header";
+        break;
+    case CURLINFO_HEADER_OUT:
+        text = "=> Send header";
+        break;
+    case CURLINFO_DATA_OUT:
+        text = "=> Send data";
+        break;
+    case CURLINFO_SSL_DATA_OUT:
+        text = "=> Send SSL data";
+        break;
+    case CURLINFO_DATA_IN:
+        text = "<= Recv data";
+        break;
+    case CURLINFO_SSL_DATA_IN:
+        text = "<= Recv SSL data";
+        break;
+    }
+    TRACE_VERBOSE("%-35s:%-8d:%-16p %s\n", __func__, __LINE__, (void*)userp, text);
+
+    return 0;
+}
+
+WINHTTPAPI HINTERNET WINAPI WinHttpOpen
+(
+    LPCTSTR pszAgentW,
+    DWORD dwAccessType,
+    LPCTSTR pszProxyW,
+    LPCTSTR pszProxyBypassW,
+    DWORD dwFlags
+)
+{
+    std::shared_ptr<WinHttpSessionImp> session = std::make_shared<WinHttpSessionImp> ();
+    if (!session)
+        return NULL;
+
+    TRACE("%-35s:%-8d:%-16p\n", __func__, __LINE__, (void*) (session.get()));
+    if (dwAccessType == WINHTTP_ACCESS_TYPE_NAMED_PROXY)
+    {
+        if (!pszProxyW)
+            return NULL;
+
+        TRACE("%-35s:%-8d:%-16p proxy: " STRING_LITERAL " \n", __func__, __LINE__, (void*) (session.get()), pszProxyW);
+        ConvertCstrAssign(pszProxyW, WCTLEN(pszProxyW), session->GetProxy());
+
+        std::vector<std::string> proxies = Split(session->GetProxy(), ';');
+        if (proxies.empty())
+        {
+            proxies.push_back(session->GetProxy());
+        }
+        session->SetProxies(proxies);
+    }
+
+    if (dwFlags & WINHTTP_FLAG_ASYNC) {
+        session->SetAsync();
+    }
+
+    WinHttpHandleContainer<WinHttpSessionImp>::Instance().Register(session);
+    return session.get();
+}
+
+WINHTTPAPI HINTERNET WINAPI WinHttpConnect
+(
+    HINTERNET hSession,
+    LPCTSTR pswzServerName,
+    INTERNET_PORT nServerPort,
+    DWORD dwReserved
+)
+{
+    WinHttpSessionImp *session = static_cast<WinHttpSessionImp *>(hSession);
+    if (!session)
+        return NULL;
+
+    TRACE("%-35s:%-8d:%-16p pswzServerName: " STRING_LITERAL " nServerPort:%d\n",
+          __func__, __LINE__, (void*)session, pswzServerName, nServerPort);
+    ConvertCstrAssign(pswzServerName, WCTLEN(pswzServerName), session->GetServerName());
+
+    session->SetServerPort(nServerPort);
+    std::shared_ptr<WinHttpConnectImp> connect = std::make_shared<WinHttpConnectImp> ();
+    if (!connect)
+        return NULL;
+
+    connect->SetHandle(session);
+    WinHttpHandleContainer<WinHttpConnectImp>::Instance().Register(connect);
+    return connect.get();
+}
+
+
+BOOLAPI WinHttpCloseHandle(
+    HINTERNET hInternet
+)
+{
+    WinHttpBase *base = static_cast<WinHttpBase *>(hInternet);
+
+  
```

Comment:
what happens if "offers" is a tls version higher than the used libcurl understands? wouldn't this line cause curl to use a sslversion < what was offered?

Reply:
That's true. Offered is not what is actually used. Upper level gives a list of options that the transport protocol can use. What actually gets used depends on the curl version. 

If there is a complete mismatch in terms of capabilities, CURL will use the CURL_SSLVERSION_DEFAULT as a minimum and CURL_SSLVERSION_MAX_DEFAULT as a maximum.

These definitions come from the current curl library in use.

Are you interested in enforcing the offered as required?